### DISCUSSION TITLE
Avalonia UI & CSLA

### DISCUSSION THREAD

**Discussion by aprokopyev** [MarimerLLC/csla#2564] [Ideas]
Hello,

What do you think about integrating Avalonia UI with CSLA ?
https://avaloniaui.net/

It is a cross-platform .NET UI framework using XAML similar to WPF.

More controls:
https://github.com/PieroCastillo/Aura.UI
https://github.com/AvaloniaCommunity/awesome-avalonia#libraries--extensions
https://github.com/topics/avaloniaui

Thanks

**Comment by rockfordlhotka** [comment]
Hi @aprokopyev - I've looked at Avalonia a couple times over the years. I think having a UI support project for it would be a fine idea - along the lines of the WPF/UWP support in the current Csla.Xaml projects.

I suppose a `Csla.Xaml.Avalonia` project that relies on the existing `Csla.Xaml.Shared` code and a `Csla.Avalonia` nuget package?

Is this something you are interested in contributing?

**Reply by Chicagoan2016** [reply]
@rockfordlhotka  as soon as I went to https://avaloniaui.net/ I saw @tfreitasleal  picture there : ).
Third picture from right, in the last row!!

**Comment by aprokopyev** [comment]
Hi Rockford,

Thank you for your welcoming Avalonia in CSLA.

I am not sure that I am the best match for this work because I have only some theoretical knowledge of your CSLA.NET project and I do not have any significant experience even with Microsoft WPF. On the other side I have a lot of experience coding VB.NET and DevExpress XtraGrid for WinForms. Also I know C# well enough just to do the same things which I can do in VB.NET (actually I coded C++ in my school times).

I need Avalonia for one of my further hobby tasks "Agreement Generator" program to make it cross platform, now it uses DevExpress XtraGrid, nHibernate and PostgreSQL. It would be very nice also to add CSLA to it, but I am not sure how much time would it take me for a first time to get all this working together like Avalonia+CSLA+nHibernate.

**Reply by rockfordlhotka** [reply]
@aprokopyev I appreciate your reply and assessment of your background.

I do not think the core of CSLA will need to change to support Avalonia. The framework is already built to work across so many different environments and UI frameworks that it will probably work just fine in Avalonia.

However, there are satellite projects for each UI framework to provide productivity helpers to make it easier to use CSLA objects from each UI framework. For example, Csla.Windows.Forms and Csla.Xaml and Csla.Blazor.

The Csla.Xaml project is relevant to Avalonia. This project supports WPF and Xamarin because (though they are slightly different) both UI frameworks rely on XAML and have similar controls and requirements.

I suspect that Avalonia will benefit from the same controls and helpers.

Some things, like `ViewModel` and `PropertyInfo` will probably work with little or no change. Other things, like `PropertyStatus`, may or may not work at all (it works in WPF, but not Xamarin).

The `Csla.Xaml` projects are just _helpers_ though. They are not required to use CSLA, but they are intended to make the UI developer's life easier when using CSLA domain objects.
